### PR TITLE
Added decorator exclude_xframe_options_header

### PR DESCRIPTION
**Pull request by jkfran**
## Done

- Added decorator `exclude_xframe_options_header` to avoid the security header `X-Frame-Options` with the value `SAMEORIGIN`

## QA

QA this change with https://github.com/canonical-web-and-design/snapcraft.io/pull/3337

### CODE REVIEW

**Comment by nottrobin** [comment]
So the interface for using this feature is:

``` python3
@exclude_xframe_options_header
def my_view():
    # ...
```

Another option would be something like:

``` python3
def my_view():
    # ...
    response = flask.render_template("template")
    response.headers["x-frame-options"] = "allowall"
    return response
```

And then we would write the `after_request` function to only define `x-frame-options` if it didn't already exist. This feels a little more intuitive to me, but it has the downside that `allowall` is in fact not a valid option for `x-frame-options` - it only works because setting `x-frame-options` to an invalid option will cause browsers to simply ignore the header, which has the same effect.

So I can't quite decide what is best here.

**Comment by jkfran** [comment]
> Another option would be something like:
> 
> ```python
> def my_view():
>     # ...
>     response = flask.render_template("template")
>     response.headers["x-frame-options"] = "allowall"
>     return response
> ```

This was my first choice until I searched what is the default value for `x-frame-options` and discovering that there isn't and the header shouldn't be set then. I don't like setting an invalid value just to avoid the issue, I like the decorator, I'm also open to new ideas.

As an improvement of the current change, we can mix these two options so if we need to set the values `DENY` or `ALLOW-FROM ...` this decorator is not needed because it's already defined.

File: canonicalwebteam/flask_base/app.py [hunk 1]
```diff
@@ -20,7 +20,17 @@
...
+
+    # Check if view_function has exclude_xframe_options_header decorator
+    if flask.request.endpoint in flask.current_app.view_functions:
+        view_func = flask.current_app.view_functions[flask.request.endpoint]
+        add_xframe_options_header = not hasattr(
+            view_func, "_exclude_xframe_options_header"
+        )
+
+    if add_xframe_options_header:
+        response.headers["X-Frame-Options"] = "SAMEORIGIN"
```
Comment (nottrobin) [inline canonicalwebteam/flask_base/app.py:33]: Could you just add an `if` here to only set the header if it's not already set? That will then allow us to override it with `DENY` or something if needed.

**Comment by nottrobin** [comment]
Yes I agree with all that. Let's do both =)

**Review by nottrobin — APPROVED** [review]
Approved, assuming you put in the conditional =)

**Comment by jkfran** [comment]
> Approved, assuming you put in the conditional =)

I did! 😃

**Pull request merged**